[BUG] App crashing because of themes resources

With the new theme pickers introduced in #481 and #482, the preference value used to remember the picked theme is derived from its resource id, but those are not fixed, so while it may work for some time it causes crashes with errors such as 


```kt
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.github.livingwithhippos.unchained.debug/com.github.livingwithhippos.unchained.base.MainActivity}

Caused by: android.view.InflateException: Binary XML file line #46 in com.github.livingwithhippos.unchained.debug:layout/activity_main: Binary XML file line #46 in com.github.livingwithhippos.unchained.debug:layout/activity_main: Error inflating class com.google.android.material.bottomnavigation.BottomNavigationView

Caused by: android.view.InflateException: Binary XML file line #46 in com.github.livingwithhippos.unchained.debug:layout/activity_main: Error inflating class com.google.android.material.bottomnavigation.BottomNavigationView

Caused by: java.lang.IllegalArgumentException: This component requires that you specify a valid TextAppearance attribute. Update your app theme to inherit from Theme.MaterialComponents (or a descendant).
```

replace with the theme key we already have which is stable instead

--- a/app/app/src/main/java/com/github/livingwithhippos/unchained/base/ThemingCallback.kt
+++ b/app/app/src/main/java/com/github/livingwithhippos/unchained/base/ThemingCallback.kt
@@ -8,14 +8,15 @@ import androidx.appcompat.app.AppCompatActivity
 import androidx.appcompat.app.AppCompatDelegate
 import com.github.livingwithhippos.unchained.R
 import androidx.core.content.ContextCompat
-import com.github.livingwithhippos.unchained.settings.view.CUSTOM_THEME_KEY
 import com.github.livingwithhippos.unchained.settings.view.SettingsFragment
+import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.KEY_NEW_THEME
 import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.THEME_AUTO
 import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.THEME_DAY
 import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.THEME_NIGHT
 import com.github.livingwithhippos.unchained.settings.view.ThemeItem
 import com.github.livingwithhippos.unchained.utilities.PreferenceKeys
 import com.github.livingwithhippos.unchained.utilities.extension.applyThemeAwareSystemBarIconColors
+import com.github.livingwithhippos.unchained.utilities.extension.getThemeItem
 import com.github.livingwithhippos.unchained.utilities.extension.getThemeList
 import com.google.android.material.color.DynamicColors
 import com.google.android.material.color.DynamicColorsOptions
@@ -31,15 +32,13 @@ class ThemingCallback(val preferences: SharedPreferences) : Application.Activity
     private val appliedThemes = WeakHashMap<Activity, Int>()
 
     override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
-        val themeRes = currentThemeRes()
-        val themesList = activity.applicationContext.getThemeList()
-        val currentTheme: ThemeItem = themesList.find { it.themeID == themeRes } ?: themesList[1]
+        val currentTheme = activity.applicationContext.getThemeItem()
         setupNightMode(currentTheme.nightMode)
         if (activity is AppCompatActivity) {
-            setCustomTheme(activity, themeRes)
+            setCustomTheme(activity, currentTheme.themeID)
             if (currentTheme.isDynamic) applyDynamicColors(activity, currentTheme)
             activity.applyThemeAwareSystemBarIconColors()
-            appliedThemes[activity] = currentThemeSignature(themeRes, currentTheme)
+            appliedThemes[activity] = themeSignature(currentTheme.themeID, currentTheme)
         }
     }
 
@@ -56,10 +55,8 @@ class ThemingCallback(val preferences: SharedPreferences) : Application.Activity
 
     override fun onActivityResumed(activity: Activity) {
         if (activity !is AppCompatActivity) return
-        val themeRes = currentThemeRes()
-        val themesList = activity.applicationContext.getThemeList()
-        val currentTheme: ThemeItem = themesList.find { it.themeID == themeRes } ?: themesList[1]
-        val themeSignature = currentThemeSignature(themeRes, currentTheme)
+        val theme = activity.applicationContext.getThemeItem()
+        val themeSignature = themeSignature(theme.themeID, theme)
         val appliedSignature = appliedThemes[activity]
         if (appliedSignature != null && appliedSignature != themeSignature) {
             // avoid re-triggering on the recreated instance's own resume
@@ -68,13 +65,7 @@ class ThemingCallback(val preferences: SharedPreferences) : Application.Activity
         }
     }
 
-    private fun currentThemeRes(): Int =
-        preferences.getInt(
-            SettingsFragment.KEY_THEME_NEW,
-            R.style.Theme_Unchained_Material3_Green_One,
-        )
-
-    private fun currentThemeSignature(themeRes: Int, theme: ThemeItem): Int =
+    private fun themeSignature(themeRes: Int, theme: ThemeItem): Int =
         if (theme.key == CUSTOM_THEME_KEY) {
             // the style id alone can't tell two different custom colors apart, since both use
             // the same DynamicCustom style; fold the actual seed color into the signature too
@@ -127,5 +118,11 @@ class ThemingCallback(val preferences: SharedPreferences) : Application.Activity
             THEME_DAY -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
             THEME_NIGHT -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
         }
+    }
+
+    companion object {
+        const val CUSTOM_THEME_KEY = "custom_theme"
+        const val DYNAMIC_WALLPAPER_KEY = "dynamic_wallpaper"
+        const val DEFAULT_THEME_KEY = "green_01"
     }
 }

--- a/app/app/src/main/java/com/github/livingwithhippos/unchained/settings/view/CustomColorPickerDialog.kt
+++ b/app/app/src/main/java/com/github/livingwithhippos/unchained/settings/view/CustomColorPickerDialog.kt
@@ -10,8 +10,12 @@ import android.widget.LinearLayout
 import androidx.core.content.edit
 import androidx.preference.PreferenceManager
 import com.github.livingwithhippos.unchained.R
+import com.github.livingwithhippos.unchained.base.ThemingCallback.Companion.CUSTOM_THEME_KEY
+import com.github.livingwithhippos.unchained.base.ThemingCallback.Companion.DEFAULT_THEME_KEY
 import com.github.livingwithhippos.unchained.customview.ColorWheelView
+import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.KEY_NEW_THEME
 import com.github.livingwithhippos.unchained.utilities.PreferenceKeys
+import com.github.livingwithhippos.unchained.utilities.extension.getThemeItem
 import com.github.livingwithhippos.unchained.utilities.extension.getThemeList
 import com.google.android.material.bottomsheet.BottomSheetBehavior
 import com.google.android.material.bottomsheet.BottomSheetDialog
@@ -69,13 +73,8 @@ class CustomColorPickerDialog : BottomSheetDialogFragment() {
             previewDrawable.setColor(colorWheel.selectedColor)
         }
 
-        val currentTheme = requireContext().getThemeList().find { it.key == CUSTOM_THEME_KEY }
-        val initialColor =
-            currentTheme?.primaryColorID
-                ?: preferences.getInt(
-                    PreferenceKeys.Ui.CUSTOM_THEME_SEED_COLOR_KEY,
-                    0xFF376B00.toInt(),
-                )
+        val currentTheme = requireContext().getThemeItem()
+        val initialColor = currentTheme.primaryColorID
         colorWheel.setColor(initialColor)
         brightnessSlider.value = colorWheel.value
         updatePreview()
@@ -108,7 +107,7 @@ class CustomColorPickerDialog : BottomSheetDialogFragment() {
             // both back to back here would read a stale value and silently fail to persist
             preferences.edit {
                 putInt(PreferenceKeys.Ui.CUSTOM_THEME_SEED_COLOR_KEY, colorWheel.selectedColor)
-                putInt(SettingsFragment.KEY_THEME_NEW, R.style.Theme_Unchained_Material3_DynamicCustom)
+                putString(KEY_NEW_THEME, CUSTOM_THEME_KEY)
             }
             dismiss()
             activity?.recreate()

--- a/app/app/src/main/java/com/github/livingwithhippos/unchained/settings/view/SettingsFragment.kt
+++ b/app/app/src/main/java/com/github/livingwithhippos/unchained/settings/view/SettingsFragment.kt
@@ -22,6 +22,7 @@ import com.github.livingwithhippos.unchained.settings.viewmodel.SettingEvent
 import com.github.livingwithhippos.unchained.settings.viewmodel.SettingsViewModel
 import com.github.livingwithhippos.unchained.utilities.FEEDBACK_URL
 import com.github.livingwithhippos.unchained.utilities.GPLV3_URL
+import com.github.livingwithhippos.unchained.utilities.extension.getThemeItem
 import com.github.livingwithhippos.unchained.utilities.extension.getThemeList
 import com.github.livingwithhippos.unchained.utilities.extension.openExternalWebPage
 import com.github.livingwithhippos.unchained.utilities.extension.showToast
@@ -150,11 +151,8 @@ class SettingsFragment : PreferenceFragmentCompat() {
             openThemePickerDialog()
             true
         }
-        val themeRes =
-            preferences.getInt(KEY_THEME_NEW, R.style.Theme_Unchained_Material3_Green_One)
-        val currentTheme: ThemeItem? =
-            requireContext().getThemeList().find { it.themeID == themeRes }
-        findPreference<Preference>("selected_theme")?.summary = currentTheme?.name
+        val theme = requireContext().getThemeItem()
+        findPreference<Preference>("selected_theme")?.summary = theme.name
     }
 
     private fun setupKodi() {
@@ -267,7 +265,7 @@ class SettingsFragment : PreferenceFragmentCompat() {
     companion object {
         // these must match the ones used in [xml/settings.xml]
         const val KEY_DAY_NIGHT = "day_night_theme"
-        const val KEY_THEME_NEW = "new_current_theme"
+        const val KEY_NEW_THEME = "current_theme_new"
         const val KEY_TORRENT_NOTIFICATIONS = "notification_torrent_key"
         const val KEY_REFERRAL_ASKED = "referral_asked_key"
         const val KEY_REFERRAL_USE = "use_referral_key"

--- a/app/app/src/main/java/com/github/livingwithhippos/unchained/settings/view/ThemePickerDialog.kt
+++ b/app/app/src/main/java/com/github/livingwithhippos/unchained/settings/view/ThemePickerDialog.kt
@@ -11,8 +11,10 @@ import androidx.recyclerview.widget.DiffUtil
 import androidx.recyclerview.widget.ListAdapter
 import androidx.recyclerview.widget.RecyclerView
 import com.github.livingwithhippos.unchained.R
+import com.github.livingwithhippos.unchained.base.ThemingCallback.Companion.CUSTOM_THEME_KEY
 import com.github.livingwithhippos.unchained.databinding.ItemThemeListBinding
 import com.github.livingwithhippos.unchained.settings.viewmodel.SettingsViewModel
+import com.github.livingwithhippos.unchained.utilities.extension.getThemeItem
 import com.github.livingwithhippos.unchained.utilities.extension.getThemeList
 import com.google.android.material.bottomsheet.BottomSheetBehavior
 import com.google.android.material.bottomsheet.BottomSheetDialog
@@ -59,10 +61,10 @@ class ThemePickerDialog : BottomSheetDialogFragment(), ThemePickListener {
         super.onViewCreated(view, savedInstanceState)
 
         val themes = requireContext().getThemeList()
-        val currentTheme = themes.find { it.themeID == viewModel.getCurrentTheme() }
+        val currentTheme = requireContext().getThemeItem()
 
         val recyclerView = view.findViewById<RecyclerView>(R.id.themeList)
-        adapter = ThemePickerAdapter(this, currentTheme?.key)
+        adapter = ThemePickerAdapter(this, currentTheme.key)
         recyclerView.adapter = adapter
         adapter.submitList(themes)
 
@@ -71,7 +73,7 @@ class ThemePickerDialog : BottomSheetDialogFragment(), ThemePickListener {
         viewModel.themeLiveData.observe(viewLifecycleOwner) {
             it.getContentIfNotHandled()?.let { theme ->
                 adapter.setSelectedKey(theme.key)
-                viewModel.applyTheme()
+                viewModel.applyTheme(theme)
                 dismiss()
                 activity?.recreate()
             }
@@ -111,9 +113,6 @@ data class ThemeItem(
     @param:ColorInt val primaryContainerColorID: Int,
     val isDynamic: Boolean = false,
 )
-
-/** [ThemeItem.key] of the user-customizable, seed-color-based dynamic theme. */
-const val CUSTOM_THEME_KEY = "custom_theme"
 
 class ThemePickerAdapter(private val listener: ThemePickListener, initialSelectedKey: String?) :
     ListAdapter<ThemeItem, ThemeViewHolder>(DiffCallback()) {

--- a/app/app/src/main/java/com/github/livingwithhippos/unchained/settings/viewmodel/SettingsViewModel.kt
+++ b/app/app/src/main/java/com/github/livingwithhippos/unchained/settings/viewmodel/SettingsViewModel.kt
@@ -8,11 +8,12 @@ import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import com.github.livingwithhippos.unchained.R
+import com.github.livingwithhippos.unchained.base.ThemingCallback.Companion.DEFAULT_THEME_KEY
 import com.github.livingwithhippos.unchained.data.local.ProtoStore
 import com.github.livingwithhippos.unchained.data.repository.HostsRepository
 import com.github.livingwithhippos.unchained.data.repository.KodiRepository
 import com.github.livingwithhippos.unchained.data.repository.PluginRepository
-import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.KEY_THEME_NEW
+import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.KEY_NEW_THEME
 import com.github.livingwithhippos.unchained.settings.view.ThemeItem
 import com.github.livingwithhippos.unchained.start.viewmodel.MainActivityViewModel.Companion.KEY_DOWNLOAD_FOLDER
 import com.github.livingwithhippos.unchained.utilities.Event
@@ -75,13 +76,12 @@ constructor(
         }
     }
 
-    fun applyTheme() {
-        val selectedTheme: ThemeItem? = themeLiveData.value?.peekContent()
-        selectedTheme?.let { preferences.edit { putInt(KEY_THEME_NEW, it.themeID) } }
+    fun applyTheme(selectedTheme: ThemeItem) {
+        preferences.edit { putString(KEY_NEW_THEME, selectedTheme.key) }
     }
 
-    fun getCurrentTheme(): Int {
-        return preferences.getInt(KEY_THEME_NEW, R.style.Theme_Unchained_Material3_Green_One)
+    fun getCurrentTheme(): String {
+        return preferences.getString(KEY_NEW_THEME, DEFAULT_THEME_KEY) ?: DEFAULT_THEME_KEY
     }
 }
 

--- a/app/app/src/main/java/com/github/livingwithhippos/unchained/utilities/extension/Extension.kt
+++ b/app/app/src/main/java/com/github/livingwithhippos/unchained/utilities/extension/Extension.kt
@@ -31,6 +31,7 @@ import android.view.WindowInsetsController
 import android.widget.Toast
 import androidx.annotation.AttrRes
 import androidx.annotation.DrawableRes
+import androidx.annotation.RequiresApi
 import androidx.appcompat.app.AppCompatActivity
 import androidx.core.content.res.ResourcesCompat
 import androidx.core.graphics.ColorUtils
@@ -39,7 +40,10 @@ import androidx.core.view.WindowInsetsControllerCompat
 import androidx.fragment.app.Fragment
 import androidx.preference.PreferenceManager
 import com.github.livingwithhippos.unchained.R
-import com.github.livingwithhippos.unchained.settings.view.CUSTOM_THEME_KEY
+import com.github.livingwithhippos.unchained.base.ThemingCallback.Companion.CUSTOM_THEME_KEY
+import com.github.livingwithhippos.unchained.base.ThemingCallback.Companion.DEFAULT_THEME_KEY
+import com.github.livingwithhippos.unchained.base.ThemingCallback.Companion.DYNAMIC_WALLPAPER_KEY
+import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.KEY_NEW_THEME
 import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.THEME_AUTO
 import com.github.livingwithhippos.unchained.settings.view.SettingsFragment.Companion.THEME_DAY
 import com.github.livingwithhippos.unchained.settings.view.ThemeItem
@@ -150,6 +154,7 @@ fun Context.getThemeList(): List<ThemeItem> {
  * color system instead of a fixed palette, so its preview swatch is only a best-effort
  * approximation of the current wallpaper colors, not the exact colors DynamicColors will apply
  */
+@RequiresApi(Build.VERSION_CODES.S)
 private fun Context.getDynamicWallpaperThemeItem(): ThemeItem {
     val isNight =
         resources.configuration.uiMode.and(Configuration.UI_MODE_NIGHT_MASK) ==
@@ -160,7 +165,7 @@ private fun Context.getDynamicWallpaperThemeItem(): ThemeItem {
     val containerTone = if (isNight) android.R.color.system_accent2_700 else android.R.color.system_accent2_100
     return ThemeItem(
         name = "Material You",
-        key = "dynamic_wallpaper",
+        key = DYNAMIC_WALLPAPER_KEY,
         nightMode = THEME_AUTO,
         themeID = R.style.Theme_Unchained_Material3_Dynamic,
         primaryColorID = ResourcesCompat.getColor(resources, primaryTone, null),
@@ -193,6 +198,14 @@ private fun Context.getCustomThemeItem(): ThemeItem {
         primaryContainerColorID = ColorUtils.blendARGB(seedColor, Color.WHITE, 0.7f),
         isDynamic = true,
     )
+}
+
+fun Context.getThemeItem(): ThemeItem {
+    val themes = this.getThemeList()
+    val preferences = PreferenceManager.getDefaultSharedPreferences(this)
+    val themeKey = preferences.getString(KEY_NEW_THEME, DEFAULT_THEME_KEY)
+    val theme = themes.find { it.key == themeKey } ?: themes[6]
+    return theme
 }
 
 /**

